# SQL 4 Data Science: Predict Project
## Bhejane Online Trading Store

### Student Starter Notebook

© Explore Data Science Academy

### Honour Code

I {**Nompumelelo**, **Ncube **}, confirm - by submitting this document - that the solutions in this notebook are a result of my own work and that I abide by the EDSA honour code (https://drive.google.com/file/d/1QDCjGZJ8-FmJE3bZdIQNwnJyQKPhHZBn/view?usp=sharing).

Non-compliance with the honour code constitutes a material breach of contract.

<a id='Context'></a>
### 1. Context

The Bhejane trading store is an online retailer specialising in Covid essential items. The store has recently been struggling with the management of its database-related inventory system. Luckily for them, you've been hired as a consultant to fix the problem. 

Its time to flex your Ninja SQL skills! 
 
**Your mission, should you choose to accept it:**

You will receive a denormalised database consisting of two tables. Your task is to transform the database such that is in third normal form (3NF). To ensure that a consistent normalisation process is followed by each student, we have attached an Entity Relationship Diagram (ERD) which is a guideline on what tables need to be produced. Once you have normalised the database, you will be required to complete an MCQ test found under the 'Predict' tab of Athena.  

Though the normalisation process, you will be guided into acknowledging client (Bhenjane Trading) use-cases for the database, and dealing with data anomalies in SQL. 

![Bhejane company logo](https://raw.githubusercontent.com/Explore-AI/Pictures/master/SQL4DS/SQL_Predict/Bhejane.png)

<div align="center" >
    Bhejane, or the Black Rhino. Image by <a href="https://commons.wikimedia.org/wiki/File:Black_Rhino_(Diceros_bicornis)_browsing_..._(46584052962).jpg">Wikimedia Commons</a>
</div>

<a id='Imports'></a>
### 2. Imports
Please use the below command to install sql_magic, this is the package that will assist you with SQL syntax hightlighting.
* pip install sql_magic

Remember to start each new cell with " %%read_sql "

In [1]:
%load_ext sql

In [2]:
%%sql 

sqlite:///bhejane.db

<a id='Data_description'></a>
### 3. Data description

The original database consist of 2 tables. 
* Product Table
* Transaction Table

The `Product` table consists of the stock of all inventory that Bhejane has on hand currently, or has had on hand historically. Items which are in this table are able to be purchased, and a record of all sales (transactions) in 2020 is notorised in the `Transactions` table. To link the tables - the `barcode` can be used. Any item in the `Transactions` table, must therefore appear in the `Products` table. 

In [3]:
# DO NOT EDIT THIS CELL
import pandas as pd
data_description = pd.read_excel('Data Description.xlsx')
data_description

Table Name         Column Name  \
0       Products               Width   
1       Products              Length   
2       Products              Height   
3       Products             Barcode   
4       Products            Quantity   
5       Products               Brand   
6       Products      NavigationPath   
7       Products              Colour   
8       Products        StockCountry   
9       Products  ProductDescription   
10      Products            PackType   
11      Products        Volume_litre   
12      Products            Warranty   
13      Products           Weight_kg   
14      Products     ItemDescription   
15      Products               Price   
16  Transactions              CartID   
17  Transactions             Barcode   
18  Transactions               Total   
19  Transactions            UserName   
20  Transactions         InvoiceDate   

                                          Desciption   
0                 Width of the product once assembled  
1                Length of the product once assembled  
2                Height of the product once assembled  
3                      The unique product identifier   
4                            Number of goods in stock  
5      Product brand name relating to product company  
6                 Navigation path to specific product  
7                 Name default colour for the product  
8             Country where the stock was bought from  
9                            Descriptive product name  
10           The way in which the product is packaged  
11              The quantity of liquid based products  
12                              Product warranty type  
13                              Weight of the product  
14            Description of contents of the package   
15                   The cost of the product in Rands  
16  Identifies an individual basket from a user pu...  
17               Barcode for item that user purchased  
18                      Total value of a transaction   
19                  Name of customer purchasing stock  
20               Date on which the product was bought

<a id='Setting_up'></a>
### 4. Setting up the database

In [5]:
import sqlite3

# Now you can use sqlite3
conn = sqlite3.connect('bhejane.db')
cursor = conn.cursor()


#### Creating both Products and Transaction Tables

In [6]:
%%sql
--DO NOT EDIT THIS CELL
DROP TABLE IF EXISTS "Products";
DROP TABLE IF EXISTS "Transactions";

CREATE TABLE "Products" (
    "Width"   REAL,
    "Length"  REAL,
    "Height"  REAL,
    "Barcode" VARCHAR(150),
    "Quantity" REAL,
    "Brand" VARCHAR(150), 
    "NavigationPath" VARCHAR(150),
    "Colour" VARCHAR(150),
    "StockCountry" VARCHAR(150),
    "ProductDescription" VARCHAR(150),
    "PackType" VARCHAR(150), 
    "Volume_litre" REAL, 
    "Warranty" VARCHAR(150), 
    "Weight_kg" REAL,
    "ItemDescription" VARCHAR(150), 
    "Price" REAL
);


CREATE TABLE "Transactions" (
    "CartID" INTEGER,
    "Barcode" VARCHAR(150), 
    "Total" REAL,
    "UserName" VARCHAR(150), 
    "InvoiceDate" DATETIME
);

 * sqlite:///bhejane.db
Done.
Done.
Done.
Done.


[]

#### Load Data into Product and Transaction tables

Although we have not yet covered python in the sprint at EXPLORE - you have additional resources at your disposal, one of which happens to be a data engineering expert, and she has created a brief script for you which can be used to extract the data from the csv files, and load it into a sqlite database. 

This will be included in your submission, so that the end-to-end processing of the data is visible, and repeatable for any additional consultants who may be brought on board at a later stage. 


In [8]:
import pandas as pd


# Establish a connection to the SQLite database
conn = sqlite3.connect('bhejane.db')

# Load data from CSV into a DataFrame
products_df = pd.read_csv('bhejane_covid_essentials_Products.csv')

# Insert data into the Products table
products_df.to_sql('Products', conn, if_exists='append', index=False)

# Load data from another CSV into a DataFrame
transactions_df = pd.read_csv('bhejane_covid_essentials_Transactions.csv')

# Insert data into the Transactions table
transactions_df.to_sql('Transactions', conn, if_exists='append', index=False)

# Close the database connection
conn.close()

<a id='ERD'></a>
### 5. Denormalized Database Tables

<br>

<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/master/SQL4DS/SQL_Predict/Denormalized_Tables.png" alt="Denormalized Tables" border="0">

#### 5.1 Explore the Denormalized Tables
Familiarize yourself with the data given in the two tables by writing your own SQL queries to explore properties of the dataset.i.e Look for data inconsistencies, anormalies, redundancies etc to guide your normalization process. 

This section will not be graded - think of it as your own scratch pad.



In [9]:
%%sql
-- #Use this section to explore the database by writing your own queries
SELECT * FROM Products
LIMIT 2;

 * sqlite:///bhejane.db
Done.


[(None, None, None, '300507946', 493.0, 'Hikvision', 'Computers & Tablets / Smart Home & Connected Living / Home Security / Security Cameras', None, None, 'Hikvision 1080P 2MP Turbo HD IR Bullet Camera', None, 0.0, 'Limited (6 months)', None, '1 x Hikvision 1080P Bullet camera', 399.0),
 (None, None, None, '300507946', 493.0, 'Hikvision', 'Computers & Tablets / Smart Home & Connected Living / Home Security / Security Cameras', None, None, 'Hikvision 1080P 2MP Turbo HD IR Bullet Camera', None, 0.0, 'Limited (6 months)', None, 'Manual', 399.0)]

### 6. Normalize the given Database tables to the 1st Normal Form (1NF)

Given the below below target ERD create new tables such the the database conforms to the 1st Normal Form

![1st Normal Form](https://raw.githubusercontent.com/Explore-AI/Pictures/master/SQL4DS/SQL_Predict/1stNF.png)


It is suggested that you create all your tables before attempting to populate them with data, this will help reduce errors that might creep up due to logical dependencies.

#### 6.1 Create the tables required for the 1st Normal Form
Use the above ERD sketch to create the required tables. Please label your tables as they appear in the ERD sketch

In [10]:
%%sql
--#Create tables required for 1NF
DROP TABLE IF EXISTS "Products_1NF";
DROP TABLE IF EXISTS "Transactions_1NF";

CREATE TABLE "Products_1NF"(
    "Barcode" VARCHAR(150),
    "NavigationPath" VARCHAR(150),
    "ItemDescription" VARCHAR(150),
    "Colour" VARCHAR(150),
    "ProductDescription" VARCHAR(150) NOT NULL,
    "Brand" VARCHAR(150),
    "Price" REAL NOT NULL,
    "Quantity" INTEGER NOT NULL,
    "PackType" VARCHAR(150),
    "Warranty" VARCHAR(150),
    "StockCountry" VARCHAR(150),
    "Weight_kg" REAL,
    "Volume_litre" REAL,
    "Length" REAL,
    "Width" REAL,
    "Height" REAL,
    PRIMARY KEY("Barcode","NavigationPath","ItemDescription")
);

CREATE TABLE "Transactions_1NF" (
    "CartID" INTEGER NOT NULL,
    "Barcode" VARCHAR(150) NOT NULL,
    "UserName" VARCHAR(150) NOT NULL,
    "InvoiceDate" DATETIME NOT NULL,
    "Total" REAL NOT NULL,
    FOREIGN KEY ("Barcode") REFERENCES "Products_1NF" ("Barcode"),
    PRIMARY KEY("CartID", "Barcode", "UserName")
);

 * sqlite:///bhejane.db
Done.
Done.
Done.
Done.


[]

#### 6.2 Populate the tables you have create in the above section.

Populate the tables such that the database conforms to the 1st Normal Form

In [11]:
%%sql
--#Populate the 1NF tables
DELETE FROM "Products_1NF";
DELETE FROM "Transactions_1NF";

INSERT INTO "Products_1NF" ("Barcode","NavigationPath","ItemDescription","Colour","ProductDescription","Brand",
                            "Price","Quantity","PackType","Warranty","StockCountry","Weight_kg",
                            "Volume_litre","Length","Width","Height")
SELECT DISTINCT
     Barcode
    ,NavigationPath
    ,ItemDescription
    ,Colour
    ,ProductDescription
    ,Brand
    ,Price
    ,Quantity
    ,PackType
    ,Warranty
    ,StockCountry
    ,Weight_kg
    ,Volume_litre
    ,Length
    ,Width
    ,Height 
FROM 
    Products;

INSERT INTO "Transactions_1NF"("CartID","Barcode","UserName","InvoiceDate","Total")
SELECT DISTINCT 
    CartID
    ,Barcode
    ,UserName
    ,InvoiceDate
    ,Total
FROM
    Transactions;
    

 * sqlite:///bhejane.db
Done.
0 rows affected.
1162 rows affected.
278 rows affected.


[]

###  7. Convert the database into its 2nd Normal Form (2NF).

Please label your tables as they appear in the ERD sketch.

#### 7.1 Entity Relationship Diagram

![2nd Normal Form](https://raw.githubusercontent.com/Explore-AI/Pictures/master/SQL4DS/SQL_Predict/2ndNF.png)

#### 7.2 2NF Requirements
To transition from 1NF to 2NF, we need all columns in all tables to not have a partial dependancy on the PK of the table. This means that any tables which have a composite key e.g. `PRIMARY KEY("Barcode","NavigationPath","ItemDescription")` cannot have columns in the table which are dependant on only `Barcode`, `NavigationPath`, or `ItemDescription`. 

In moving to 2NF we will seperate out each of these as a primary key (for their own respective tables), and any columns dependant on one of these columns but not the others, will go into that table. The only columns remaining in the tables will be fully functionally dependant on the primary key of the table. For example, 

Thereby, we remove the partial dependancy, and will be in 2NF. We are instructed to create a unique identifier for the `Transactions_2NF`. 

**Second normal form notes**

<br>$\bullet$ Notice that `Total` is only dependant on `CartID` and not on `<CartID,barcode>`. This is a partial dependancy
<br>$\bullet$ Second normal form required no *partial functional dependancy* on the key. 
<br>$\bullet$ This can be achieved in 2 ways: create a unique identifier per row (so that there is a simple primary key), or break the table up into only the columns which are fully dependant on the composite key. 
<br>$\bullet$ We are going to opt for the easier solution, and create a unique simple primary key on the `Transactions_1NF`. We can create a unique primary key by concatenating `CartID||barcode`
<br>$\bullet$ NB - this approach does make it more challenging to get into 3NF! Which would be a good argument to opt for the latter approach. 

#### 7.3 Data Anomalies

You may stumble into an error when you insert the data which violates this constraint. You query this with the data owner, and you are told that there are three transactions in the database which were errors. You are informed that *all transactions in the database correspond to products are in the database*.

These entries were captured incorrectly, and must be removed. How will you deal with this? 

In [12]:
%%sql
SELECT COUNT(DISTINCT barcode) FROM Transactions_1NF WHERE barcode NOT IN (SELECT barcode FROM products_1NF);

 * sqlite:///bhejane.db
Done.


[(2,)]

### <font color='turquoise'>Action: remove the affected entries with a `delete` query </font>

In [13]:
%%sql
delete from  Transactions_1NF
WHERE CartID < 0;

 * sqlite:///bhejane.db
0 rows affected.


[]

In [14]:
%%sql
-- Should show no entries!
select count(distinct barcode) from Transactions_1NF where barcode not in (select barcode from products_1NF)

 * sqlite:///bhejane.db
Done.


[(2,)]

#### 7.4 FK Requirements

_Note: sqlite does not always enforce foreign key constraints (and it is even more inconsistent with magic commands to enforce FK constraints!), so it is critical that you understand the purpose of the foreign key constraint and therefore are able to enforce the FK constraint yourself!_

To enforce data integrity, and to ensure that strange things cannot happen - we will have a FK constraint. A foreign key forces all values of the FK, to be limited to only the values which exist as entries in the PK column of the table for which the aforementioned table is a foreign key to. For example, we cannot sell a product (in our Transactions table) which we do not have in our Products table. 

You will need to work around this by removing the erroneous data from the tables before continuing, since this would violate the FK constraint between the Transactions_2NF and Products_2NF table otherwise. 

_Hint: Identify the observations in the Transactions_1NF table, and remove them from there._

#### 7.5 Client Use-Case Requirements

The data owner has advised that there were several instances of incorrect data capturing. Whenever data was partially captured, it was redone, but the partially-complete entries have not been removed from the database. 

For all tables which are going to be used, consider the table utility relative to the Products (Products_2NF) table. You are tasked with removing 'useless' entries as you take the data from 1NF to 2NF. For the sake of ease-of-use for the client, we will removing all rows in the 2NF tables `PackageContents_2NF` and `Colours_2NF` which have an empty value for the attributes that the client will use them for - these are the `ItemDescription` and `Colour` columns, respectively. 

_Hint: Make inserts into the other 2NF (`PackageContents_2NF` and `Colours_2NF`) tables using a `where` clause to remove instances of the value being an empty string, or a `null` value. This will need to be considered for the `ItemDescription` and for the `Colour` when data is inserted into the corresponding tables above (i.e. only insert data into the `PackageContents_2NF` which has a value for the `ItemDescription`, and only insert data into the `Colours_2NF` table which has a value for `Colour`_

### <font color='turquoise'>Action: Investigate these anomalies for the tables mentioned </font>

In [15]:
%%sql 

SELECT DISTINCT
    ItemDescription,
    PackType,
    Warranty
FROM
    Products
WHERE ItemDescription = '';

 * sqlite:///bhejane.db
Done.


[]

### Let's do it! Construct the database above in second normal form

In [16]:
%%sql
--#Create tables required for 2NF

DROP TABLE IF EXISTS "Products_2NF";
DROP TABLE IF EXISTS "Transactions_2NF";
DROP TABLE IF EXISTS "Navigation_2NF";
DROP TABLE IF EXISTS "PackageContents_2NF";
DROP TABLE IF EXISTS "Colours_2NF";

CREATE TABLE Products_2NF(
    "RegistryID" INTEGER NOT NULL,
    "Barcode" VARCHAR(150) NOT NULL,
    "ProductDescription" VARCHAR(150) NOT NULL,
    "Brand" VARCHAR(150),
    "Price" REAL NOT NULL,
    "Quantity" INTEGER NOT NULL,
    "StockCountry" VARCHAR(150),
    "Weight_kg" REAL,
    "Volume_litre" REAL,
    "Length" REAL,
    "Width" REAL,
    "Height" REAL,
    "PathID" INTEGER,
    "ItemID" INTEGER,
    "ColourID" INTEGER,
    PRIMARY KEY(RegistryID AUTOINCREMENT),
    CONSTRAINT fk_nav FOREIGN KEY('PathID') REFERENCES Navigation_2NF ('PathID'),
    CONSTRAINT fk_pack FOREIGN KEY('ItemID') REFERENCES PackageContents_2NF ('ItemID'),
    CONSTRAINT fk_col FOREIGN KEY('ColourID') REFERENCES Colours_2NF ('ColourID') 
);

CREATE TABLE "Transactions_2NF" (
    "CartID||Barcode||UserName" VARCHAR(150) NOT NULL,
    "CartID" INTEGER NOT NULL,
    "Barcode" VARCHAR(150) NOT NULL,
    "UserName" VARCHAR(150) NOT NULL,
    "InvoiceDate" DATETIME NOT NULL,
    "Total" REAL NOT NULL,
    PRIMARY KEY("CartID||Barcode||UserName")
    CONSTRAINT fk_trans FOREIGN KEY('Barcode') REFERENCES Products_2NF ('Barcode')
);


CREATE TABLE "Navigation_2NF"(
    "PathID" INTEGER NOT NULL,
    "NavigationPath" VARCHAR(150),
    PRIMARY KEY("PathID" AUTOINCREMENT)
);

CREATE TABLE "Colours_2NF"(
    "ColourID" INTEGER NOT NULL,
    "Colour" VARCHAR(150),
    PRIMARY KEY("ColourID" AUTOINCREMENT)
);

CREATE TABLE "PackageContents_2NF"(
    "ItemID" INTEGER NOT NULL,
    "ItemDescription" VARCHAR(150),
    "PackType" VARCHAR(150),
    "Warranty" VARCHAR(150),
    PRIMARY KEY("ItemID" AUTOINCREMENT)
);


 * sqlite:///bhejane.db
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.


[]

### Let's do it! _Populate_ the database above in second normal form

Firstly, let's consider the entries which are `null` and `= ''`. And we will only insert the relevant entries. 

In [17]:
%%sql
-- #Populate the tables so that they conform to 2NF

DELETE FROM "Products_2NF";
DELETE FROM "Transactions_2NF";
DELETE FROM "Navigation_2NF";
DELETE FROM "PackageContents_2NF";
DELETE FROM "Colours_2NF";


INSERT INTO "Navigation_2NF"("NavigationPath")
SELECT DISTINCT 
    NavigationPath 
FROM
    Products_1NF;
    
INSERT INTO "PackageContents_2NF"("ItemDescription","PackType","Warranty")
SELECT DISTINCT 
    ItemDescription,
    PackType,
    Warranty 
FROM
    Products_1NF
WHERE ItemDescription <> ''
        ;
    
INSERT INTO "Colours_2NF"("Colour")
SELECT DISTINCT 
    Colour
FROM
    Products_1NF
WHERE Colour <> ''
        ;

 * sqlite:///bhejane.db
Done.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
396 rows affected.
600 rows affected.
17 rows affected.


[]

A more complicated scenario below as we insert into `Products_2NF`. We will insert from the Products_1NF table. We can use LEFT JOIN, because the data in each of these tables that `Products_1NF` has a FK to, originally came from `Products_1NF`, so we are gauranteed to get matches back. Normally however, what you would get from the FK restriction is the requirement to use _parent-table_ `LEFT JOIN` _child-table_. 

In [18]:
%%sql 

INSERT INTO "Products_2NF" ("PathID","ItemID","ColourID","Barcode","ProductDescription","Brand","Price",
                            "Quantity","StockCountry","Weight_kg","Volume_litre",
                            "Length","Width","Height")
SELECT
    PathID,
    ItemID,
    ColourID,
    Products_1NF.Barcode,
    ProductDescription,
    Brand,
    Price,
    Quantity,
    StockCountry,
    Weight_kg, 
    Volume_litre, 
    Length,
    Width,
    Height 
FROM 
    Products_1NF
LEFT JOIN Navigation_2NF ON Products_1NF.NavigationPath = Navigation_2NF.NavigationPath 
LEFT JOIN PackageContents_2NF ON Products_1NF.ItemDescription = PackageContents_2NF.ItemDescription
LEFT JOIN Colours_2NF ON Products_1NF.Colour = Colours_2NF.Colour

 * sqlite:///bhejane.db
1214 rows affected.


[]

Last but not least, we insert into the `Transactions_2NF`table. Recall that we created a new PK in this table, and pay attention to how we are inserting values into the table. 

In [19]:
%%sql 

INSERT INTO "Transactions_2NF"("CartID||Barcode||UserName","CartID","Barcode","UserName","InvoiceDate","Total")
SELECT DISTINCT 
    CartID||Transactions_1NF.Barcode||UserName
    ,CartID
    ,Transactions_1NF.Barcode
    ,UserName
    ,InvoiceDate
    ,Total
FROM
    Transactions_1NF
LEFT JOIN Products_2NF ON Transactions_1NF.Barcode = Products_2NF.Barcode ;


 * sqlite:///bhejane.db
278 rows affected.


[]

<a id='Target_ERD'></a>
### 8. Convert the table into its 3rd Normal Form (3NF)

Please label your tables as they appear in the ERD sketch. In the same fashion as done in 2NF, note which tables need to have data inserted `where column <> '' and column is not null`.

<br>

![3rd Normal Form](https://raw.githubusercontent.com/Explore-AI/Pictures/master/SQL4DS/SQL_Predict/3rdNF.png)

In [20]:
%%sql

--#Create tables required for 2NF
DROP TABLE IF EXISTS "Transactions_3NF";
DROP TABLE IF EXISTS "Carts_3NF";
DROP TABLE IF EXISTS "Products_3NF";
DROP TABLE IF EXISTS "Users_3NF";
DROP TABLE IF EXISTS "Navigation_3NF";
DROP TABLE IF EXISTS "PackageContents_3NF";
DROP TABLE IF EXISTS "Colours_3NF";
DROP TABLE IF EXISTS "Brands_3NF";
DROP TABLE IF EXISTS "Locations_3NF";

CREATE TABLE "Products_3NF" (
    "RegistryID" INTEGER NOT NULL,
    "Barcode" VARCHAR(150) NOT NULL,
    "ProductDescription" VARCHAR(150) NOT NULL,
    "Price" REAL NOT NULL,
    "Quantity" INTEGER NOT NULL,
    "Weight_kg" REAL,
    "Volume_litre" REAL,
    "Length" REAL,
    "Width" REAL,
    "Height" REAL,
    "PathID" INTEGER,
    "ItemID" INTEGER,
    "ColourID" INTEGER,
    "BrandID" INTEGER,
    "LocationID" INTEGER,
    PRIMARY KEY("RegistryID" AUTOINCREMENT),
    FOREIGN KEY("PathID") REFERENCES "Navigation_3NF" ("PathID"),
    FOREIGN KEY("ItemID") REFERENCES "PackageContents_3NF" ("ItemID"),
    FOREIGN KEY("ColourID") REFERENCES "Colours_3NF" ("ColourID") ,
    FOREIGN KEY ("BrandID") REFERENCES "Brands_3NF" ("BrandID"),
    FOREIGN KEY ("LocationID") REFERENCES "Locations_3NF" ("LocationID")
);


CREATE TABLE "Carts_3NF" (
    "CartID" INTEGER NOT NULL,
    "InvoiceDate" DATETIME NOT NULL,
    "Total" REAL NOT NULL,
    PRIMARY KEY ("CartID")
);


CREATE TABLE "Users_3NF"(
    "UserID" INTEGER NOT NULL,
    UserName VARCHAR(50) NOT NULL,
    PRIMARY KEY("UserID" AUTOINCREMENT)
);


 * sqlite:///bhejane.db
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.


[]

### <font color='turquoise'>Action: proceed with the construction of the remaining tables mentioned </font>

In [21]:
%%sql

CREATE TABLE "Transactions_3NF" (
    "CartID" VARCHAR(150) NOT NULL,
    "Barcode" VARCHAR(150) NOT NULL,
    "Username" VARCHAR(150) NOT NULL,
    "UserID" INTEGER NOT NULL,
    PRIMARY KEY ("CartID", "Barcode", "Username"),
    FOREIGN KEY ("UserID") REFERENCES "Users_3NF" ("UserID"),
    FOREIGN KEY ("Barcode") REFERENCES "Products_3NF" ("Barcode")
);
CREATE TABLE "Navigation_3NF" (
    "PathID" INTEGER NOT NULL,
    "NavigationPath" VARCHAR(150),
    PRIMARY KEY ("PathID")
);

CREATE TABLE "PackageContents_3NF" (
    "ItemID" INTEGER NOT NULL,
    "ItemDescription" VARCHAR(150) NOT NULL,
    "PackType" VARCHAR(150) NOT NULL,
    "Warranty" VARCHAR(150) NOT NULL,
    PRIMARY KEY ("ItemID")
); 

CREATE TABLE "Colours_3NF" (
    "ColourID" INTEGER NOT NULL,
    "Colour" VARCHAR(150) NOT NULL,
    PRIMARY KEY ("ColourID")
);

CREATE TABLE "Brands_3NF" (
    "BrandID" INTEGER NOT NULL,
    "Brand" VARCHAR(150) NOT NULL,
    PRIMARY KEY ("BrandID")
);

CREATE TABLE "Locations_3NF" (
    "LocationID" INTEGER NOT NULL,
    "StockCountry" VARCHAR(150) NOT NULL,
    PRIMARY KEY ("LocationID")
);

 * sqlite:///bhejane.db
Done.
Done.
Done.
Done.
Done.
Done.


[]

In [22]:
%%sql

--# Populate the tables to that they conform to the 3rd Normal Form
DELETE FROM "Products_3NF";
DELETE FROM "Transactions_3NF";
DELETE FROM "Carts_3NF";
DELETE FROM "Users_3NF";
DELETE FROM "Navigation_3NF";
DELETE FROM "PackageContents_3NF";
DELETE FROM "Colours_3NF";
DELETE FROM "Brands_3NF";
DELETE FROM "Locations_3NF";

INSERT INTO "Users_3NF" ("UserName")
SELECT DISTINCT UserName FROM Transactions_2NF;


 * sqlite:///bhejane.db
Done.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
160 rows affected.


[]

### <font color='turquoise'>Action: proceed with the insertion into `Carts_3NF`.   </font>

Take note that there are duplicate values coming from `Transactions_2NF` - retain the `distinct` combinations of `CartID`,`InvoiceDate`,`Total` only.

In [23]:
%%sql
INSERT INTO Carts_3NF (CartID, InvoiceDate, Total)
SELECT DISTINCT CartID, InvoiceDate, Total
FROM Transactions_2NF;


 * sqlite:///bhejane.db
192 rows affected.


[]

### <font color='turquoise'>Action: proceed with the insertion into the tables below. </font>

Similarly to the table above, mind the occurances of duplicate entries. These can be avoided by using `SELECT distinct` to avoid inserting redundant rows. 


In [24]:
%%sql

INSERT INTO Locations_3NF (StockCountry)
SELECT COALESCE("StockCountry", 'Unknown')
FROM Products_2NF;

INSERT INTO Brands_3NF (Brand)
SELECT COALESCE("Brand", 'Unknown')
FROM Products_2NF;

INSERT INTO "Colours_3NF" ("Colour")
SELECT DISTINCT "Colour"
FROM Products_2NF;

INSERT INTO "PackageContents_3NF" ("ItemDescription", "PackType", "Warranty")
SELECT DISTINCT "ItemDescription", "PackType", "Warranty"
FROM Products_2NF;

INSERT INTO "Navigation_3NF" ("NavigationPath")
SELECT DISTINCT "NavigationPath"
FROM Products_2NF;


 * sqlite:///bhejane.db
1214 rows affected.
1214 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.


[]

### <font color='turquoise'>Action: proceed with the insertion into `Products_3NF`. </font>
Take note of the joins required for additional data contained in other tables. 

In [25]:
%%sql
INSERT INTO "Products_3NF" ("PathID", "ItemID", "ColourID", "BrandID", "LocationID",
                            "Barcode", "ProductDescription", "Price",
                            "Quantity", "Weight_kg", "Volume_litre",
                            "Length", "Width", "Height")
SELECT
    N.PathID,
    PC.ItemID,
    C.ColourID,
    B.BrandID,
    L.LocationID,
    P.Barcode,
    P.ProductDescription,
    P.Price,
    P.Quantity,
    P.Weight_kg,
    P.Volume_litre,
    P.Length,
    P.Width,
    P.Height
FROM Navigation_3NF AS N
JOIN Products_3NF AS P ON P.PathID = N.PathID
JOIN PackageContents_3NF AS PC ON P.ItemID = PC.ItemID
JOIN Colours_3NF AS C ON P.ColourID = C.ColourID
JOIN Brands_3NF AS B ON P.BrandID = B.BrandID
JOIN Locations_3NF AS L ON P.LocationID = L.LocationID

 * sqlite:///bhejane.db
0 rows affected.


[]

### <font color='turquoise'>Action: proceed with the insertion into `Transactions_3NF`.   </font>

Follow previous proceedure, and checks. 

In [26]:
%%sql
INSERT INTO "Transactions_3NF" ("CartID", "Barcode", "UserName", "UserID")
SELECT DISTINCT "CartID", "Barcode", "UserName", "UserID"
FROM Transactions_2NF;


 * sqlite:///bhejane.db
278 rows affected.


[]

### <font color='turquoise'>Action: Investigate the DB.   </font>

Note - the tables counts below are what you should arrive at. If you don't, reconsider how you have done the inserts above. Having the tables correctly formatted will allow to answer the questions more easily, and without errors.

| Table Name | Count |
| --- | --- |
| Brands_3NF | 232 |
| Carts_3NF |  190 |
| Colours_3NF | 17 |
| Locations_3NF | 2 |
| Navigation_3NF | 396 |
| PackageContents_3NF | 600 |
| Products_3NF | 1214 |
| Transactions_3NF | 275 |
| Users_3NF | 158 |

In [27]:
%%sql 

-- How many entries are here in each of the tables in the database now? 
select 'Products_3NF' as table_name, count(*) from Products_3NF
UNION
select 'Transactions_3NF' as table_name, count(*) from Transactions_3NF
UNION
select 'Users_3NF' as table_name, count(*) from Users_3NF
UNION
select 'Navigation_3NF' as table_name, count(*) from Navigation_3NF
UNION
select 'PackageContents_3NF' as table_name, count(*) from PackageContents_3NF
UNION
select 'Colours_3NF' as table_name, count(*) from Colours_3NF
UNION
select 'Brands_3NF' as table_name, count(*) from Brands_3NF
UNION
select 'Locations_3NF' as table_name, count(*) from Locations_3NF
UNION
select 'Carts_3NF' as table_name, count(*) from Carts_3NF


 * sqlite:///bhejane.db
Done.


[('Brands_3NF', 1214),
 ('Carts_3NF', 192),
 ('Colours_3NF', 1),
 ('Locations_3NF', 1214),
 ('Navigation_3NF', 1),
 ('PackageContents_3NF', 1),
 ('Products_3NF', 0),
 ('Transactions_3NF', 278),
 ('Users_3NF', 160)]

<a id='MCQ_questions'></a>
## 9. MCQ Questions

Having completed the normalisation of the database, you may use the following cells to help you answer the MCQs for the predict. 

Q1) How many unique products does the company have?

In [28]:
%%sql

SELECT ROUND(SUM(quantity), 0) AS TotalStock
FROM Products;

 * sqlite:///bhejane.db
Done.


[(615016.0,)]

In [29]:
%%sql 
select *
from products



 * sqlite:///bhejane.db
Done.


[(None, None, None, '300507946', 493.0, 'Hikvision', 'Computers & Tablets / Smart Home & Connected Living / Home Security / Security Cameras', None, None, 'Hikvision 1080P 2MP Turbo HD IR Bullet Camera', None, 0.0, 'Limited (6 months)', None, '1 x Hikvision 1080P Bullet camera', 399.0),
 (None, None, None, '300507946', 493.0, 'Hikvision', 'Computers & Tablets / Smart Home & Connected Living / Home Security / Security Cameras', None, None, 'Hikvision 1080P 2MP Turbo HD IR Bullet Camera', None, 0.0, 'Limited (6 months)', None, 'Manual', 399.0),
 (None, None, None, '10325354918', 467.0, 'ZEE', 'Fashion / Accessories / Scarves', 'Grey', None, 'ZEE 3-in-1 Unisex Gaiter', None, 0.0, 'Limited (6 months)', None, None, 139.0),
 (None, None, None, '27131187035', 275.0, 'Estee Lauder', 'Beauty / Luxury Beauty / Makeup / Face / Foundation', 'Fresco', 'South Africa', 'Estee Lauder Double Wear Stay In Place Makeup', None, 0.0, 'Non-Returnable', None, None, 655.0),
 (None, None, None, '27131187035', 275.0, 'Estee Lauder', 'Beauty / Luxury Beauty / Shop By Brand / Estee Lauder / Makeup / Face / Foundation', 'Fresco', 'South Africa', 'Estee Lauder Double Wear Stay In Place Makeup', None, 0.0, 'Non-Returnable', None, None, 655.0),
 (None, None, None, '27131187035', 275.0, 'Estee Lauder', 'Beauty / Makeup / Face / Foundation', 'Fresco', 'South Africa', 'Estee Lauder Double Wear Stay In Place Makeup', None, 0.0, 'Non-Returnable', None, None, 655.0),
 (None, None, None, '39800085139', 247.0, 'Energizer', 'Cameras / Cameras & Lenses / Camera Accessories / General Purpose Batteries', None, 'South Africa', 'Energizer 3V CR2450 Lithium Coin Battery', None, 0.0, 'Limited (12 months)', None, None, 49.0),
 (None, None, None, '43859499182', 386.0, 'Fellowes', 'Office & Stationery / Stationery / Office Supplies / Cleaning', None, 'South Africa', 'Fellowes HFC Free Air Duster - 350ml', None, 0.0, 'Limited (6 months)', None, '1 x 350ml Free Air Duster', 149.0),
 (None, None, None, '43917915548', 338.0, 'WAHL', 'Health / Personal Care / Shaving & Hair Removal / Shaving Equipment & Devices / Body Groomers', None, 'South Africa', 'Wahl Home Pro Basic Corded 8 Piece Haircutting Kit', None, 0.0, 'Limited (6 months)', None, 'Wahl Home Pro Basic Corded 8 Piece Haircutting Kit', 149.0),
 (None, None, None, '43917915548', 338.0, 'WAHL', 'Health / Personal Care / Shaving & Hair Removal / Shaving Equipment & Devices / Shavers', None, 'South Africa', 'Wahl Home Pro Basic Corded 8 Piece Haircutting Kit', None, 0.0, 'Limited (6 months)', None, 'Wahl Home Pro Basic Corded 8 Piece Haircutting Kit', 149.0),
 (None, None, None, '43917915548', 338.0, 'WAHL', 'Beauty / Hair Care / Hair Styling Accessories / Clips & Barrettes', None, 'South Africa', 'Wahl Home Pro Basic Corded 8 Piece Haircutting Kit', None, 0.0, 'Limited (6 months)', None, 'Wahl Home Pro Basic Corded 8 Piece Haircutting Kit', 149.0),
 (None, None, None, '43917915548', 338.0, 'WAHL', 'Health / Personal Care / Shaving & Hair Removal / Hair Removal', None, 'South Africa', 'Wahl Home Pro Basic Corded 8 Piece Haircutting Kit', None, 0.0, 'Limited (6 months)', None, 'Wahl Home Pro Basic Corded 8 Piece Haircutting Kit', 149.0),
 (None, None, None, '43917915548', 338.0, 'WAHL', 'Beauty / Hair Care / Hair Styling Tools / Clippers', None, 'South Africa', 'Wahl Home Pro Basic Corded 8 Piece Haircutting Kit', None, 0.0, 'Limited (6 months)', None, 'Wahl Home Pro Basic Corded 8 Piece Haircutting Kit', 149.0),
 (None, None, None, '43917963358', 375.0, 'WAHL', 'Beauty / Hair Care / Hair Styling Accessories / Clips & Barrettes', 'Red', 'South Africa', 'Wahl Easy Cut 15 Piece Complete Hair Clipper Kit', None, 0.0, 'Limited (6 months)', None, 'Wahl Easy Cut 15 Piece Complete Hair Clipper Kit', 199.0),
 (None, None, None, '43917963358', 375.0, 'WAHL', 'Beauty / Hair Care / Hair Styling Tools / Clippers', 'Red', 'South Africa', 'Wahl Easy Cut 15 Piece Complete Hair Clipper Kit', None, 0.0, 'Limited (6 months)', None, 'Wahl Easy Cu

Q2) How many users bought from Bhejane in April 2020?

In [30]:
%%sql
SELECT UserName
FROM Transactions
WHERE InvoiceDate = '2020-03-25'

 * sqlite:///bhejane.db
Done.


[]

Q3) How many users bought 3 or more items that cost more than R1000?

In [31]:
%%sql
SELECT DISTINCT InvoiceDate
FROM Transactions
WHERE UserName = 'Mandla';


 * sqlite:///bhejane.db
Done.


[('2020-06-07 0:00:00',), ('2020-04-09 0:00:00',), ('2020-04-11 0:00:00',)]

Q4) Which user made the largest purchase on a single transaction?

In [32]:
%%sql
SELECT Brand, COUNT(DISTINCT ProductDescription) AS Variety
FROM Products
GROUP BY Brand
ORDER BY Variety DESC
LIMIT 5;

 * sqlite:///bhejane.db
Done.


[(None, 180),
 ('SanDisk', 16),
 ('Pampers', 11),
 ('Russell Hobbs', 10),
 ('Canon', 9)]

Q5) How many components does the product: "5m Colour Changing RGB LED Strip Light" (MPTAL57588104) come with?

In [33]:
%%sql
SELECT Total
FROM Transactions
WHERE UserName = 'Zanele' AND InvoiceDate = '2020-04-04';

 * sqlite:///bhejane.db
Done.


[]

Q6) How many brands are available at Bhejane?

In [34]:
%%sql
SELECT UserName, SUM(Total) AS TotalSpent
FROM Transactions
WHERE InvoiceDate >= '2020-01-01' AND InvoiceDate <= '2020-12-31'
GROUP BY UserName
HAVING COUNT(*) = 4
ORDER BY TotalSpent DESC
LIMIT 1;

 * sqlite:///bhejane.db
Done.


[('keith', 2980.0)]

7) What is the price of the "Verimark - Floorwiz 2in1 Mop"?

In [35]:
%%sql
SELECT SUM(Transactions.Total) AS TotalPrice
FROM Transactions
JOIN Products ON Transactions.Barcode = Products.Barcode
WHERE Products.NavigationPath IN ('Estee Lauder Double Wear Stay In Place Makeup', 'Angelcare - Nappy Bin Refill - 3 Pack', 'RCT 650VA Line Interactive UPS')

 * sqlite:///bhejane.db
Done.


[(None,)]

Q8) Calculate the package volume of the "Russell Hobbs - Slow Cooker" using the given dimensions

In [36]:
%%sql
SELECT ProductDescription, Quantity
FROM Products
ORDER BY Quantity DESC
LIMIT 1;

 * sqlite:///bhejane.db
Done.


[('Beard Growth Essential Oil', 500.0)]

Q9) Which user made the most transactions in the Year 2020?

In [37]:
%%sql
SELECT COUNT(*) AS NumProducts
FROM Products
WHERE StockCountry = 'United Kingdom';


 * sqlite:///bhejane.db
Done.


[(8,)]

Q10) What is the total number of users that shop at Bhejane?

In [38]:
%%sql
SELECT ProductDescription, Colour
FROM Products
WHERE Barcode = 'TAL00035388021';

 * sqlite:///bhejane.db
Done.


[('Angelcare - Nappy Bin Refill - 3 Pack', 'Blue'),
 ('Angelcare - Nappy Bin Refill - 3 Pack', 'Blue'),
 ('Angelcare - Nappy Bin Refill - 3 Pack', 'Blue'),
 ('Angelcare - Nappy Bin Refill - 3 Pack', 'Blue')]

Q11) What is the record count for the Colours_3NF Table?

In [39]:
%%sql
SELECT COUNT(DISTINCT Brand) AS BeerBrandCount
FROM products
WHERE Brand IN (
    'Three Ships', 'Wellington', 'Richelieu', 'Klipdrift', 'Metalix',
    'Castle Lite', 'Castle Free', 'Heineken'
    
);

 * sqlite:///bhejane.db
Done.


[(8,)]

Q12) What would the total price be if I had the following items in my cart?
* MPTAL57588104
* 5000394203921
* 6932391917652

In [40]:
%%sql
SELECT ProductDescription, COUNT(DISTINCT NavigationPath) AS PathCount, Quantity
FROM Products
GROUP BY ProductDescription, Quantity
HAVING PathCount > 2
ORDER BY Quantity DESC
LIMIT 10;


 * sqlite:///bhejane.db
Done.


[('Croxley Create Fine Nib Whiteboard Markers - Blister of 10', 11, 482.0),
 ('Addis - Window Squeegee Long Handle', 3, 474.0),
 ("Butterfly Filing Pockets A4 100's", 5, 471.0),
 ('Art Set - 66 Piece', 4, 468.0),
 ('Pilot G-2 0,5 Fine Nib Gel Retractable Pen - Black', 6, 442.0),
 ('Wifi Extender / Repeater - up to 300mbps', 3, 439.0),
 ('Pritt Stick 43g Glue Stick', 6, 431.0),
 ('100% Cotton 3-ply Face Mask - Navy', 3, 422.0),
 ('SanDisk 64GB 98 MB/s Ultra Micro UHS-l SDXC C10', 3, 419.0),
 ('Energizer CR2032 Lithium Coin Battery - 3V', 3, 418.0)]

Q13) What is the barcode of the most sold product?

In [41]:
%%sql
SELECT Transactions.UserName, Products.Warranty
FROM Transactions
JOIN Products ON Transactions.Barcode = Products.Barcode
WHERE Products.ProductDescription = 'Canon PG-445 & CL-446 Ink Cartridges Multipack'
  AND strftime('%Y-%m', Transactions.InvoiceDate) = '2020-05'
LIMIT 1;


 * sqlite:///bhejane.db
Done.


[]

Q14) What are the products of Cornelis’ cart on the 2020-06-28 0:00:00?

In [42]:
%%sql
SELECT DISTINCT Transactions.UserName
FROM Transactions
JOIN Products ON Transactions.Barcode = Products.Barcode
WHERE Products.Brand = 'Russell Hobbs'
  AND Products.Colour = 'Silver';


 * sqlite:///bhejane.db
Done.


[('Karyn',), ('Clodeane',)]

Q15) Which users bought locally produced, black-coloured products on odd-numbered months of the year, and what was the total cost (rounded to the nearest integer) of the carts containing these products?

In [43]:
%%sql

SELECT SUM(Price) AS TotalPrice
FROM Products
WHERE ProductDescription IN ('Estee Lauder Double Wear Stay In Place Makeup', 'Angelcare - Nappy Bin Refill - 3 Pack', 'RCT 650VA Line Interactive UPS');


 * sqlite:///bhejane.db
Done.


[(6356.0,)]